RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h

```diff
@@ -0,0 +1,306 @@
+#ifndef RecoLocalTracker_SiPixelClusterizer_plugins_gpuClustering_h
+#define RecoLocalTracker_SiPixelClusterizer_plugins_gpuClustering_h
+
+#include <cstdint>
+#include <cstdio>
+
+#include "Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/HistoContainer.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cuda_assert.h"
+
+#include "gpuClusteringConstants.h"
+
+namespace gpuClustering {
+
+#ifdef GPU_DEBUG
+  __device__ uint32_t gMaxHit = 0;
+#endif
+
+  __global__ void countModules(uint16_t const* __restrict__ id,
+                               uint32_t* __restrict__ moduleStart,
+                               int32_t* __restrict__ clusterId,
+                               int numElements) {
+    int first = blockDim.x * blockIdx.x + threadIdx.x;
+    for (int i = first; i < numElements; i += gridDim.x * blockDim.x) {
+      clusterId[i] = i;
+      if (InvId == id[i])
+        continue;
+      auto j = i - 1;
+      while (j >= 0 and id[j] == InvId)
+        --j;
+      if (j < 0 or id[j] != id[i]) {
+        // boundary...
+        auto loc = atomicInc(moduleStart, MaxNumModules);
+        moduleStart[loc + 1] = i;
+      }
+    }
+  }
+
+  __global__
+      //  __launch_bounds__(256,4)
```

Comment:
@slava77 thank you very much for the thorough review, it's much appreciated.

I can try and take care of some requests centrally:
  - renaming files and classes
  - removing obsolete headers
  - renaming variables
  - migrating to ES consumes (though I may need help from @makortel for this)
  - use the autogenerated cfi files when they are available
  - check why PixelVertexMonitor_cff.py ended up here and in case remove it

About GPU-specific named constants, I would leave it to @VinInn and @makortel to comment what the numbers mean.
Then, I can either implement the changes, or they can make a PR to the Patatrack repository

Some things I would leave to the DPG, @tsusa @mmusich and @ferencek; I would suggest to address them in a follow up PR:
  - reduce the duplication of code and constants
  - address changes related to the differences between Phase 2 and Phase 1
  - review the hard coded limitations (at least, I would need to understand what is required globally and what can be configured at compile or runtime)

I don't know anything about: only the pre-splitting part is defined, what about the setup for after splitting?

For the usefulness or removal of `SiPixelDigiHeterogeneousConverter`, I would defer to @makortel .

That leaves the printf in the kernels; as far as I know there is no way to interface it with the MessageLogger, other than asking the MessageLogger what level is enabled for a given module, converting it into a "debug" flag, and passing it to the kernels. However I'm not sure if this is the way we want to go ?

Reply:
This is a [possible optimisation](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#launch-bounds), but it's likely to be device dependent, and it looks like we never uncommented it in the Patatrack branch.
So I think it can be removed.

Reply:
Done in https://github.com/cms-patatrack/cmssw/pull/593/commits/b7c587960302b89e417aad3df6bf00294bb888e2 .